source/common/upstream/leds.h

```diff
@@ -0,0 +1,89 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/config/endpoint/v3/endpoint_components.pb.h"
+#include "envoy/config/endpoint/v3/endpoint_components.pb.validate.h"
+#include "envoy/config/subscription.h"
+#include "envoy/local_info/local_info.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "source/common/config/subscription_base.h"
+#include "source/common/upstream/upstream_impl.h"
+
+namespace Envoy {
+namespace Upstream {
+
+/**
+ * All per-cluster LEDS stats. @see stats_macros.h
+ * These will be added to the subscription stats.
+ */
+#define ALL_LEDS_STATS(COUNTER) COUNTER(update_empty)
+
+/**
+ * Struct definition for all per-cluster LEDS stats. @see stats_macros.h
+ */
+struct LedsStats {
+  ALL_LEDS_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+/*
+ * A single subscription for all LEDS resources of a specific SourceConfig that
+ * fetches updates from a Locality Endpoint Discovery Service.
+ * Multiple subscriptions with the same LEDS collection name can use a single
+ * subscription.
+ */
+class LedsSubscription
+    : private Envoy::Config::SubscriptionBase<envoy::config::endpoint::v3::LbEndpoint>,
+      private Logger::Loggable<Logger::Id::upstream> {
+public:
+  using UpdateCb = std::function<void()>;
+  using LbEndpointList = std::list<envoy::config::endpoint::v3::LbEndpoint>;
+
+  LedsSubscription(const envoy::config::endpoint::v3::LedsClusterLocalityConfig& leds_config,
+                   const std::string& cluster_name,
+                   Server::Configuration::TransportSocketFactoryContext& factory_context,
+                   Stats::Scope& stats_scope, const UpdateCb& callback);
+
+  // Fetch all the endpoints in the locality subscription.
+  const LbEndpointList& getEndpoints() const { return endpoints_; }
```

Comment:
Done.
The idea to only return the list was to abstract over the resource names, as they are not really needed by the consumer, but it may be relevant for future changes, so modified to returning the map instead.